File: rubyonrails/notejam/test/controllers/users_controller_test.rb
require 'test_helper'

class UsersControllerTest < ActionController::TestCase
  test 'should successfully signup' do
    assert_difference('User.count') do
      post :signup, user: {
        email: 'user@example.com', password: 'password',
        password_confirmation: 'password'
      }
    end
    assert_redirected_to signin_path
  end

  test 'should not signup if invalid email' do
    post :signup, user: {
      email: 'invalid email', password: 'password',
      password_confirmation: 'password'
    }
    assert_select '.errorlist li', 'email is invalid'
  end

  test 'should not signup if required fields are missing' do
    post :signup, user: {
      email: '', password: '',
      password_confirmation: ''
    }
    assert_select '.errorlist li'
  end

  test 'should not signup if email exists' do
    email = users(:existent_user).email
    post :signup, user: {
      email: email, password: 'password',
      password_confirmation: 'password'
    }
    assert_select '.errorlist li', 'email has already been taken'
  end

  test 'should not signup if passwords do not match' do
    post :signup, user: {
      email: 'user@example.com', password: 'password1',
      password_confirmation: 'password2'
    }
    assert_select '.errorlist li'
  end

  test 'should successfully signin' do
    email = users(:existent_user).email
    post :signin,
         email: email, password: 'secure_password'

    assert_redirected_to notes_path
  end

  test 'should not signin if wrong email/password' do
    post :signin,
         email: 'some_email@example.com', password: 'password'

    assert_select '.alert-error', 'Invalid email or password'
  end
end

File: cakephp/notejam/tests/TestCase/Controller/UsersControllerTest.php
<?php
namespace App\Test\TestCase\Controller;

use App\Controller\UsersController;
use Cake\ORM\TableRegistry;
use Cake\TestSuite\IntegrationTestCase;

/**
 * App\Controller\UsersController Test Case
 */
class UsersControllerTest extends IntegrationTestCase
{
    /**
     * Fixtures
     *
     * @var array
     */
    public $fixtures = [
        'app.users'
    ];

    /**
     * Test success signup
     *
     * @return void
     */
    public function testSignupSuccess()
    {
        $data = [
            'email' => 'user@example.com',
            'password' => 'pa$$word',
            'confirm_password' => 'pa$$word'
        ];
        $this->post('/signup', $data);
        $this->assertResponseSuccess();
        $users = TableRegistry::get('Users');
        $query = $users->find()->where(['email' => $data['email']]);
        $this->assertEquals(1, $query->count());
    }

    /**
     * Test if signup fails if required fields are missing
     *
     * @return void
     */
    public function testSignupFailRequiredFields()
    {
        $data = [
            'email' => '',
            'password' => '',
            'confirm_password' => ''
        ];
        $this->post('/signup', $data);
        $this->assertResponseSuccess();
        $this->assertResponseContains('This field cannot be left empty');
    }

    /**
     * Test if signup fails if email is invalid
     *
     * @return void
     */
    public function testSignupFailInvalidEmail()
    {
        $data = [
            'email' => 'invalid email'
        ];
        $this->post('/signup', $data);
        $this->assertResponseSuccess();
        $this->assertResponseContains('The provided value is invalid');
    }

    /**
     * Test if signup fails if email already exists
     *
     * @return void
     */
    public function testSignupFailEmailExists()
    {
        $data = [
            'email' => 'user1@example.com',
            'password' => 'pa$$word',
            'confirm_password' => 'pa$$word'
        ];
        $this->post('/signup', $data);
        $this->assertResponseSuccess();
        $this->assertResponseContains('This value is already in use');
    }

    /**
     * Test if signup fails if passwords do not match
     *
     * @return void
     */
    public function testSignupFailPasswordsNotMatch()
    {
        $data = [
            'email' => 'user1@example.com',
            'password' => 'pa$$word1',
            'confirm_password' => 'pa$$word2'
        ];
        $this->post('/signup', $data);
        $this->assertResponseSuccess();
        $this->assertResponseContains('Passwords do not match');
    }

    /**
     * Test if signin success
     *
     * @return void
     */
    public function testSigninSuccess()
    {
        $data = [
            'email' => 'user1@example.com',
            'password' => '111111'
        ];
        $this->post('/signin', $data);
        $this->assertResponseSuccess();
        $this->assertRedirect(['controller' => 'Notes', 'action' => 'index']);
        $this->assertSession('user1@example.com', 'Auth.User.email');

    }

    /**
     * Test if signin fails if provided credentials are wroing
     *
     * @return void
     */
    public function testSigninFailWrongCredentials()
    {
        $data = [
            'email' => 'user2@example.com',
            'password' => 'wrong password'
        ];
        $this->post('/signin', $data);
        $this->assertResponseContains('Your username or password is incorrect.');
    }
}
